### ISSUE TITLE
How to avoid requesting/using microphone (receive only)

### ISSUE THREAD

**Issue by TheRealNate** [react-native-webrtc/react-native-webrtc#1304]
Hi there!

I'm trying to use this package in order to view the stream of nest cameras from the Nest SDM API. This API allows viewing these streams using receive only WebRTC.

I was previously having an issue with codecs, but that was fixed by a PR. I'm able to view and hear the video stream properly! However, in order for this to happen the app seems to have to request microphone permissions and start using the microphone (I see the orange privacy indicator dot while the feature is in use).

How can I receive only without capturing the microphone or needing the permission? I never call getUserMedia, so I'm not sure why the capture is taking place. How can I prevent this? If I don't add a NS key for microphone permission, I get an error telling me to do so.

And here is my code:

      const connection = new RTCPeerConnection({iceServers: []});

      connection.createDataChannel('dataSendChannel');

      const offer = (await connection.createOffer({
        offerToReceiveAudio: true,
        offerToReceiveVideo: true,
      })) as RTCSessionDescription;

      connection.addEventListener('track', e => {
        const event = e as unknown as {
          transceiver: RTCRtpTransceiver;
          streams: MediaStream[];
          track: MediaStreamTrack;
        };
        if (event.track.kind === 'video') {
          console.log(e);
          // We render the streams with RTCView
          setStreams(s => [...s, ...event.streams]);
        }
      });

      if (offer.sdp != null) {
        await connection.setLocalDescription(offer);

        console.log(offer.sdp);
        void callTheNestAPIWithMyOfferAndGetBackAnAnswerSDP()
          .then(res => {
            void connection.setRemoteDescription({
              sdp: res.answerSdp,
              type: 'answer',
            });
          });
      }

Here is my answer SDP:

````
v=0
o=- 0 2 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE 0 2 1
a=msid-semantic: *****
a=ice-lite
m=audio 19305 UDP/TLS/RTP/SAVPF 111
c=IN IP4 *****
a=rtcp:9 IN IP4 0.0.0.0
a=candidate: *****
a=candidate: *****
a=candidate: *****
a=candidate: *****
a=candidate: *****
a=candidate: *****
a=ice-ufrag:*****
a=ice-pwd:*****
a=fingerprint:sha-256 *****
a=setup:passive
a=mid:0
a=extmap:1 urn:ietf:params:rtp-hdrext:ssrc-audio-level
a=extmap:3 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=sendrecv
a=msid:virtual-6666 virtual-6666
a=rtcp-mux
a=rtpmap:111 opus/48000/2
a=rtcp-fb:111 transport-cc
a=fmtp:111 minptime=10;useinbandfec=1
a=ssrc:6666 cname:6666
m=video 9 UDP/TLS/RTP/SAVPF 96 97 98 99
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:*****
a=ice-pwd:*****
a=fingerprint:sha-256 *****
a=setup:passive
a=mid:1
a=extmap:2 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=extmap:13 urn:3gpp:video-orientation
a=extmap:3 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=sendrecv
a=msid:*****
a=rtcp-mux
a=rtpmap:96 H264/90000
a=rtcp-fb:96 transport-cc
a=rtcp-fb:96 ccm fir
a=rtcp-fb:96 nack
a=rtcp-fb:96 nack pli
a=rtcp-fb:96 goog-remb
a=fmtp:96 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=*****
a=rtpmap:97 rtx/90000
a=fmtp:97 apt=96
a=rtpmap:98 H264/90000
a=rtcp-fb:98 transport-cc
a=rtcp-fb:98 ccm fir
a=rtcp-fb:98 nack
a=rtcp-fb:98 nack pli
a=rtcp-fb:98 goog-remb
a=fmtp:98 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=*****
a=rtpmap:99 rtx/90000
a=fmtp:99 apt=98
a=ssrc-group:FID ***** *****
a=ssrc:***** cname:*****
a=ssrc:***** cname:*****
m=application 9 DTLS/SCTP 5000
c=IN IP4 0.0.0.0
a=ice-ufrag:*****
a=ice-pwd:*****
a=fingerprint:sha-256 *****
a=setup:passive
a=mid:2
a=sctpmap:5000 webrtc-datachannel 1024
````

**Comment by mattdinthehouse** [comment]
Hi @saghul just checking if this bug has been fixed before I attempt it? If it's not fixed yet can you also provide any notes you think would be helpful before I start :)

**Comment by saghul** [comment]
It's not fixed and it would require patching WebRTC, and I won't accept floating patches.

IIRC @danjenkins found it can be done with a custom audio device implementation, but I have no details.

**Comment by mattdinthehouse** [comment]
Thanks @saghul I'll see if I can find an existing bug report in upstream WebRTC otherwise I'll report one and hopefully get a patch merged

@danjenkins if you've got any code or notes for that workaround that'd be amazing :) cos I've either gotta fix the bug or work around it lol

**Comment by danjenkins** [comment]
Take https://github.com/mstyura/RTCAudioDevice and put it into your project, initialise the webrtc project in AppDelegate with something like this in your `didFinishLaunchingWithOptions`:

Of course, remove the extra logging

```
  WebRTCModuleOptions *options = [WebRTCModuleOptions sharedInstance];
  options.loggingSeverity = RTCLoggingSeverityInfo;
  
  id<RTCAudioDevice> device;

  device = [[AUAudioUnitRTCAudioDevice alloc] init];
  
  options.audioDevice = device;
```

That'll initialise libwebrtc with the custom audio device in that project

**Comment by Spharian** [comment]
@danjenkins is there any way to load this from React Native? Then deleting the iOS folder wont impact the workflow.

I also noticed than when I am using this fork from LiveKit I do not encounter the issue (but causes other strange issues like staying on the last frame when the streamer disconnect): https://github.com/livekit/react-native-webrtc/

**Comment by danjenkins** [comment]
The livekit fork has the patch that @saghul mentioned to libwebrtc in it... 

You don't ever want to completely delete your iOS directory... there will always be things to add to your appdelegate file

**Comment by mattdinthehouse** [comment]
> Take https://github.com/mstyura/RTCAudioDevice and put it into your project

Legend, thanks @danjenkins

**Comment by marc0kh** [comment]
@mattdinthehouse Were you able to get this to work? I am trying to find a bit clearer instructions on how to implement this

**Comment by mattdinthehouse** [comment]
@marc0kh cos I'm using Expo with audio only I ended up trying the @livekit/react-native and @livekit/react-native-webrtc packages that Spharian mentioned first which solved the problem for me so didn't end up using the RTCAudioDevice that Dan linked

The LiveKit fork has all the same WebRTC objects available in the normal react-native-webrtc package and doesn't require me to use any of LiveKit's cloud services of the "room" system so my code moved over without any changes :)

**Comment by luigbren** [comment]
> @marc0kh cos I'm using Expo with audio only I ended up trying the @livekit/react-native and @livekit/react-native-webrtc packages that Spharian mentioned first which solved the problem for me so didn't end up using the RTCAudioDevice that Dan linked
> 
> The LiveKit fork has all the same WebRTC objects available in the normal react-native-webrtc package and doesn't require me to use any of LiveKit's cloud services or the "room" system they've built so my code moved over without any changes :)

hello friend, please can you tell me how did you make the @livekit/react-native-webrtc package work ? because in my component I used to do the import:
```js
 import {
  RTCPeerConnection,
  RTCIceCandidate,
  RTCSessionDescription,
  RTCView,
  mediaDevices,
  NativeModules,
} from 'react-native-webrtc';
```
and after I uninstalled react-native-webrtc I proceeded to install: npm install --save @livekit/react-native-webrtc and then I do the import in the same way:
```js
 import {
  RTCPeerConnection,
  RTCIceCandidate,
  RTCSessionDescription,
  RTCView,
  mediaDevices,
  NativeModules,
} from '@livekit/react-native-webrtc';
```
but in my component the camera is not seen, it is black, but when I import 'react-native-webrtc' there if the camera is seen, I don't understand

**Comment by luigbren** [comment]
> Take https://github.com/mstyura/RTCAudioDevice and put it into your project, initialise the webrtc project in AppDelegate with something like this in your `didFinishLaunchingWithOptions`:
> 
> Of course, remove the extra logging
> 
> ```
>   WebRTCModuleOptions *options = [WebRTCModuleOptions sharedInstance];
>   options.loggingSeverity = RTCLoggingSeverityInfo;
>   
>   id<RTCAudioDevice> device;
> 
>   device = [[AUAudioUnitRTCAudioDevice alloc] init];
>   
>   options.audioDevice = device;
> ```
> 
> That'll initialise libwebrtc with the custom audio device in that project

can you give me more details please, how can I implement this patch in my react-native project, do I have to clone the whole project or just copy some files, I am confused, I need help, I need to disable the microphone when I create a response to just watch and listen to the stream.

**Comment by TadasSink** [comment]
> Take https://github.com/mstyura/RTCAudioDevice and put it into your project, initialise the webrtc project in AppDelegate with something like this in your `didFinishLaunchingWithOptions`:
> 
> Of course, remove the extra logging
> 
> ```
>   WebRTCModuleOptions *options = [WebRTCModuleOptions sharedInstance];
>   options.loggingSeverity = RTCLoggingSeverityInfo;
>   
>   id<RTCAudioDevice> device;
> 
>   device = [[AUAudioUnitRTCAudioDevice alloc] init];
>   
>   options.audioDevice = device;
> ```
> 
> That'll initialise libwebrtc with the custom audio device in that project

@saghul @danjenkins I tried assigning `AUAudioUnitRTCAudioDevice` as in the snippet you provided, but it seems that data is not flowing to `AUAudioUnit` through `RTCAudioDeviceDelegate` - buffer seems to always be empty, so I cant't hear audio. On my app I have peer connection with couple of transceivers for receiving audio and using the latest `react-native-webrtc: 124.0.4`.
 
`AUAudioUnitRTCAudioDevice` initialises without any errors. When no audio device is assigned and default audio device is used everything works as expected and I can hear audio but with mic being enabled. When running sample app from the https://github.com/mstyura/RTCAudioDevice project everything also seems to be fine.  I also tried playing with `AVAudioSession` but nothing seems to have an effect. Maybe you have an idea what I am missing?

**Comment by saghul** [comment]
You might need to activate the audio session by hand.

**Comment by TadasSink** [comment]
@saghul Thank you for a quick response. In my case It wasn't needed to activate session by hand, but this did the trick:
```
RTCAudioSessionConfiguration* config = [RTCAudioSessionConfiguration webRTCConfiguration];

AVAudioSession * session = [AVAudioSession sharedInstance];
[session setCategory:config.category withOptions:config.categoryOptions error:nil];
```

Also it seems I cannot change use `AVAudioSession` to `AVAudioSessionCategoryPlayback`. Is this limitation caused by audio device's use of `AUAudioUnit`?

**Comment by saghul** [comment]
I haven't used that myself, sorry, it was a shot in the dark 😅

**Comment by luigbren** [comment]
Please can someone give more details of the implementation?  https://github.com/mstyura/RTCAudioDevice , I also need to patch the microphone when not requested, I do not understand why they have not yet updated something so important.

**Comment by TadasSink** [comment]
@luigbren If you have  non `expo` `react-native` project you have to to add this snippet to `didFinishLaunchingWithOptions` in `AppDelegate` class, and copy required swift files from https://github.com/mstyura/RTCAudioDevice to your xcode project:

```
RTCAudioSessionConfiguration* config = [RTCAudioSessionConfiguration webRTCConfiguration];

AVAudioSession * session = [AVAudioSession sharedInstance];
[session setCategory:config.category withOptions:config.categoryOptions error:nil];

id<RTCAudioDevice> device;
device = [[AUAudioUnitRTCAudioDevice alloc] init];

WebRTCModuleOptions *options = [WebRTCModuleOptions sharedInstance];
options.loggingSeverity = RTCLoggingSeverityWarning;
options.audioDevice = device;
```

If you are using expo you are out of luck an will have to wait until guys from react-native-webrtc will provide us with JS/TS interface.

**Comment by saghul** [comment]
Please do your part. You need AUAudioUnitRTCAudioDevice, so look at what files it uses and take those.

**Comment by saghul** [comment]
Looks like a linking problem. Are all pods installed?

**Comment by luigbren** [comment]
> Looks like a linking problem. Are all pods installed?

Yes!, in package.json `"react-native-webrtc": "^124.0.4"`,  and the apply: cd ios > pod install, I also saw in a comment to add in my Podfile, and I have this: 

 `pod 'react-native-webrtc', :path => '../node_modules/react-native-webrtc'`

please tell me if it is or is not necessary to have it in Podfile, to remove it from there

**Comment by saghul** [comment]
You shouldn't need to add anything to your Podfile, RN should link it automagically. Very weird that it complains!

**Comment by luigbren** [comment]
> You shouldn't need to add anything to your Podfile, RN should link it automagically. Very weird that it complains!

well, I removed react-native-webrtc from my pod file, because rn automatically links it when installing via npn, but it still doesn't recognize it, must be some missing import.

AppDelegate.mm:25:3 Unknown type name 'WebRTCModuleOptions'.
AppDelegate.mm:25:35 Use of undeclared identifier 'WebRTCModuleOptions'.

**Comment by saghul** [comment]
Oh hold on! You need to include <react-native-webrtc/react-bative-webrtc.h>

**Comment by luigbren** [comment]
> Oh hold on! You need to include <react-native-webrtc/react-bative-webrtc.h>

file not found...!! but i include: `#import <react-native-webrtc/WebRTCModuleOptions.h>` and the error disappeared, now I just have to try to compile and see if it works. thanks for the suggestion

**Comment by luigbren** [comment]
Well I did several tests without closing the app because it restarts the configuration and I summarize, if the patch works but at the same time not ....

1. I started in Android a Transmission, in iOs enter as Receiver audio is heard and do not activate the microphone, perfect, in iOs I leave to see the transmission and then re-enter, the audio of the Sender remained disabled in iOs, you can not hear.

2. Then if from iOs I start a transmission and enter Android as Receiver, the microphone is activated and in Android you can hear it.


 Then in iOs you can no longer hear the audio when you enter as Receiver, the only way is to close the app and reopen it, it only works as it should be the first time.

@saghul @TadasSink

**Comment by saghul** [comment]
As I said, I haven't used this code so I cannot help you there, sorry.

**Comment by TadasSink** [comment]
@saghul In my case I use the code only for receiving and it works every time. You'll have to debug added code in Xcode to find the problem yourself. I would probably start looking in `AudioSessionHandler`.

**Comment by luigbren** [comment]
> @luigbren In my case I use the code only for receiving and it works every time. You'll have to debug added code in Xcode to find the problem yourself. I would probably start looking in `AudioSessionHandler`.

thanks for replying...
- Can you post your example code, how do you connect to the Peer and only use this code to listen?
- Do you use InCallManager to listen on the speaker?

This is my code:

```js
const handleOffer = (id, description) => {
    peerConnection.current = new RTCPeerConnection({
      iceServers: [
        {urls: 'stun:stun.l.google.com:19302'},
        {urls: 'stun:stun1.l.google.com:19302'},
        {urls: 'stun:stun2.l.google.com:19302'},
      ],
    });

    peerConnection.current.ontrack = event => {
      setRemoteStream(event.streams[0]);
    };

    peerConnection.current.onicecandidate = event => {
      if (event.candidate) {
        socket.emit(...);
      }
    };

    peerConnection.current
      .setRemoteDescription(new RTCSessionDescription(description))
      .then(() => peerConnection.current.createAnswer())
      .then(sdp => peerConnection.current.setLocalDescription(sdp))
      .then(() => {
        socket.emit(...);
      })
      .catch(error => {
        console.error('Error handling offer:', error);
      });
  };
```

in this way I manage the offer when I receive it from the Streaming to create the connection with the Peer, I did all the steps to implement the code, is it necessary to add some code to the AppDelegate.h file, it only works once, when I enter a transmission it is heard and the microphone does not turn on, but when I leave the transmission and enter again to see a transmission, the audio is not heard and of course the microphone does not turn on.

@TadasSink

**Comment by TadasSink** [comment]
@luigbren 
- Sorry cannot not share my code regarding webRTC connection but i doubt, It has do do something with your problem as mic is heard first time you connect
- No I don't

Sorry cannot help you more you'll have solve it yourself.

**Comment by rossmartin** [comment]
@TadasSink Thank you for sharing your experience and what worked for you with this. Did you had to modify any of the swift code in https://github.com/mstyura/RTCAudioDevice ?

I've been able to add it to my project and my `AppDelegate` is able to resolve everything with this code in `didFinishLaunchingWithOptions`

```objective-c
  RTCAudioSessionConfiguration* config = [RTCAudioSessionConfiguration webRTCConfiguration];

  AVAudioSession * session = [AVAudioSession sharedInstance];
  [session setCategory:config.category withOptions:config.categoryOptions error:nil];

  id<RTCAudioDevice> device;
  device = [[AUAudioUnitRTCAudioDevice alloc] init];

  WebRTCModuleOptions *options = [WebRTCModuleOptions sharedInstance];
  options.audioDevice = device;
```

Unfortunately my project won't build because of some obscure errors in `AUAudioUnitRTCAudioDevice.swift` (screenshots below). I have some similar errors in other files withinin `RTCAudioDevice` and I feel like I'm missing something simple here. 

<img width="1146" alt="Screenshot 2025-01-14 at 9 33 48 PM" src="https://github.com/user-attachments/assets/108da7fa-d904-49f4-aeee-cea62207f3ed" />

<img width="1149" alt="Screenshot 2025-01-14 at 9 34 06 PM" src="https://github.com/user-attachments/assets/ba32eced-9e8a-4f10-b08b-ed37f56b71cf" />

My project has some existing modules with swift code so there is a bridging header. I also made sure the "Defines Modules" project option is set to "YES" and I import the `$name-swift.h` in `AppDelegate`. If it helps any this project is on RN 0.75.4 and the swift version is set to 5 in Xcode.

CC @luigbren

**Comment by TadasSink** [comment]
@rossmartin Yes, I did some minor modifications, but did not need to change lines you showed. For me it worked with no changes in the file, I just removed lines of code which were not needed for my project. Most likely it has something to do with iOS target version  you have. On my project I use 14.0, same RN and Swift version as you do. If that does help, try to build and run https://github.com/mstyura/RTCAudioDevice as standalone project and if successful see what differs between the Xcode projects. I hope this helps.
